From e01cbf021ae5bed7141e87569ba136f4065dbf5e Mon Sep 17 00:00:00 2001
From: Gareth Davidson <gaz@bitplane.net>
Date: Sun, 6 Sep 2026 16:57:05 +0100
Subject: [PATCH 4/7] Emit relocatable TempleOS modules and raw boot images

Carry AOT options through nested compilation, resolve assembly expressions and symbol imports, and retain initialized code and data with their relocations. Emit raw assembly without executable wrappers or padding, and write complete TempleOS import and termination records.

These changes form the module writer: expression resolution, symbol ownership and relocation output share the same data structures. Native module loading remains available when TempleOS headers redefine the output signature.
---
 Src/AIWNIOS_CodeGen.HC |  64 ++++++++++---
 Src/AIWNIOS_PrsExp.HC  | 103 ++++++++++++++++----
 Src/AsmResolve.HC      | 155 +++++++++++++++++++++++++++++-
 Src/AsmX64.HC          | 209 +++++++++++++++++++++++++++++++++--------
 Src/CExts.HC           |   1 +
 Src/CMain.HC           | 164 ++++++++++++++++++++++++++++----
 Src/KLoadARM.HC        |  23 ++++-
 Src/KernelA.HH         |  15 ++-
 Src/PrsStmt.HC         |  21 +++--
 Src/PrsVar.HC          |  14 +--
 c/aiwn_lexparser.h     |   2 +
 c/arm_backend.c        |   8 +-
 c/main.c               |   6 ++
 c/parser.c             |   4 +
 c/riscv_backend.c      |   8 +-
 c/x86_64_backend.c     |  13 ++-
 16 files changed, 695 insertions(+), 115 deletions(-)

diff --git a/Src/AIWNIOS_CodeGen.HC b/Src/AIWNIOS_CodeGen.HC
index a52e495..57f4d4c 100644
--- a/Src/AIWNIOS_CodeGen.HC
+++ b/Src/AIWNIOS_CodeGen.HC
@@ -4,6 +4,7 @@
 //  your backend again when you port to another arch
 //
 #ifdef IMPORT_AIWNIOS_SYMS
+import U0 __HC_CmpCtrl_SetRawAOTAsm(U8 *);
 import U0 __HC_SetAOTRelocBeforeRIP(U8 *,I64); //See aiwn.h(TAKES number of bytes before current RIP the symbol is located at)
 import U8 *__HC_CmpCtrlDel(U8 *);
 import U8 *__HC_ICAdd_Sqr(U8 *);
@@ -570,6 +571,17 @@ Bool Globalify(U8 *addr,CAddShortRelocData *data) {
   }
   return FALSE;
 }
+U8 *AOTImportName(CHash *h) {
+  CHashSrcSym *src;
+  if(h->type&HTG_SRC_SYM) {
+    src=h;
+    if(src->import_nameim)
+      return src->import_nameim;
+    if(src->import_name)
+      return src->import_name;
+  }
+  return h->str;
+}
 U0 AddShortReloc(U8 *addr,CAddShortRelocData *data) {
   CAOTImportExport *tmpaie=CAlloc(sizeof(CAOTImportExport));
   #ifdef TARGET_RISCV
@@ -577,7 +589,7 @@ U0 AddShortReloc(U8 *addr,CAddShortRelocData *data) {
   #else
   tmpaie->type=IET_REL_I32;
   #endif
-  tmpaie->str=StrNew(data->cm->h->str);
+  tmpaie->str=StrNew(AOTImportName(data->cm->h));
   tmpaie->rip=data->cc->aotc->rip+(addr-data->machine_code);
   
   tmpaie->offset=data->cm->offset;
@@ -1320,6 +1332,8 @@ next_ic:
 U8 *AiwniosCompile(CCmpCtrl *cc,I64 *res_sz=NULL,CDbgInfo **info,Bool byte_code=FALSE) {
   CRPN *head=&cc->coc.coc_head,*cur,*new;
   CCodeMisc *misc,**table;
+  Bool relocate_glbl;
+  I64 relocation_value;
   CMemberLst *mlst,*mlst2;
   COptMemberVar *mv=NULL;
   CHashImport *h;
@@ -1407,6 +1421,8 @@ skip_convert:;
   #endif
   if(cc->flags&CCF_AOT_COMPILE)
     __HC_CmpCtrl_SetAOT(acc);
+  if(cc->flags&CCF_RAW_AOT_ASM)
+    __HC_CmpCtrl_SetRawAOTAsm(acc);
   if(info) {
     idx=cc->max_line-cc->min_line;
     *info=MAlloc((idx+2)*8+sizeof(CDbgInfo));
@@ -1907,7 +1923,8 @@ break;case IC_RIP:
           h(CHashFun*)->exe_addr
         );
       }
-    } else if(!(h->type&HTF_RESOLVE)) {
+    } else if (cc->flags&CCF_AOT_COMPILE || !(h->type&HTF_RESOLVE) ||
+          h->type&HTF_UNRESOLVED) {
         new=__HC_ICAdd_Reloc(
             acc,
             cc2,
@@ -1923,9 +1940,7 @@ break;case IC_RIP:
 	);
     }
     if(cc->flags&CCF_AOT_COMPILE) {
-      if(h->type&HTT_GLBL_VAR&&h->type&HTF_EXPORT) {
-        __HC_SetAOTRelocBeforeRIP(new,h(CHashGlblVar*)->data_addr_rip-cc->aotc->rip);
-      } else if(h->type&HTT_FUN&&!Bt(&(h(CHashFun*)->flags),Cf_EXTERN)) {
+      if(h->type&HTT_FUN&&!Bt(&(h(CHashFun*)->flags),Cf_EXTERN)) {
         __HC_SetAOTRelocBeforeRIP(new,h(CHashFun*)->exe_addr-cc->aotc->rip);
       }
     }
@@ -2185,6 +2200,8 @@ break;case IC_RIP:
 			   if(mlst->flags&MLF_STATIC) {
 				 if(mlst->static_data) {
 					MemCpy(tmpgh->data+mlst->offset,mlst->static_data,mlst->member_class->size*mlst->dim.total_cnt);
+					Free(mlst->static_data);
+					mlst->static_data=NULL;
 				 }
 			   }
 			   mlst=mlst->next;
@@ -2218,21 +2235,40 @@ break;case IC_RIP:
           }else if(IsTLSReloc(misc)) {
             tmpaie=CAlloc(sizeof(CAOTImportExport));
             tmpaie->type=IET_IMM_I64;
-            tmpaie->str=StrNew(misc->h->str);
+            tmpaie->str=StrNew(AOTImportName(misc->h));
             tmpaie->rip=cc->aotc->rip+(misc->addr(U8*)-machine_code);
             tmpaie->offset=misc->offset;
             QueIns(tmpaie,cc->aot->last_ie);
           } else if(misc->h->type&HTT_GLBL_VAR){
+            if (misc->h(CHashGlblVar*)->flags&GVF_EXTERN)
+              goto norm_reloc;
             fill_user_data.cc=cc;
 	    fill_user_data.cm=misc;
 	    fill_user_data.machine_code=machine_code;
-	    if(!Globalify(misc->addr,&fill_user_data))
-		  goto norm_reloc;
+            if (!Globalify(misc->addr,&fill_user_data)) {
+              relocate_glbl=FALSE;
+              if (!(misc->h(CHashGlblVar*)->flags&GVF_EXTERN))
+                relocate_glbl=TRUE;
+              if (misc->h(CHashGlblVar*)->flags&GVF_ALIAS)
+                relocate_glbl=TRUE;
+              if (Bt(&cc->opts,OPTf_RAW_AOT) && relocate_glbl) {
+                relocation_value=misc->h(CHashGlblVar*)->data_addr_rip+
+                      misc->offset;
+                WriteProtectMemCpy(misc->addr,&relocation_value,8);
+                tmpabs=CAlloc(sizeof(CAOTAbsAddr));
+                tmpabs->next=cc->aotc->abss;
+                tmpabs->type=AAT_ADD_U64;
+                tmpabs->rip=cc->aotc->rip+
+                      (misc->addr(U8*)-machine_code);
+                cc->aotc->abss=tmpabs;
+              } else
+                goto norm_reloc;
+            }
 	  } else {
 	  norm_reloc:
 	    tmpaie=CAlloc(sizeof(CAOTImportExport));
             tmpaie->type=IET_IMM_I64;
-            tmpaie->str=StrNew(misc->h->str);
+            tmpaie->str=StrNew(AOTImportName(misc->h));
             tmpaie->rip=cc->aotc->rip+(misc->addr(U8*)-machine_code);
             tmpaie->offset=misc->offset;
             QueIns(tmpaie,cc->aot->last_ie);
@@ -2243,10 +2279,11 @@ break;case IC_RIP:
           h=CAlloc(sizeof CHashImport);
           h->type=HTT_IMPORT_SYS_SYM;
           h->str=StrNew(misc->h->str);
+          foff=0;
+          WriteProtectMemCpy(misc->addr,&foff,8);
           h->module_base=machine_code;
-          //[reloc type 1][reloc offset 4][target-offset 4][name?]NULL]
-          //Nroot adds an aditional byte to specify NULL after we load the silly sauce 
-          h->module_header_entry=CAlloc(StrLen(h->str)+8+1+1);
+          //[reloc type 1][reloc offset 4][target-offset 4][name NUL]
+          h->module_header_entry=CAlloc(StrLen(h->str)+10);
           if(IsTLSReloc(misc))
 	    h->module_header_entry[0]=IET_IMM_I64;
           else 
@@ -2293,6 +2330,9 @@ break;case IC_RIP:
     if(cur->user_data2) I64SetDel(cur->user_data2);
     if(cur->type==IC_ASM) {
 //TODO AOT
+      if(cc->flags&CCF_AOT_COMPILE)
+        cur->ic_data(CAiwniosAOTBlob*)->final_start_rip=cc->aotc->rip+
+              (cur->ic_data(CAiwniosAOTBlob*)->final_start_addr-machine_code);
       JITFixupAiwniosAOTBlob(cc,cur->ic_data,cur->ic_data(CAiwniosAOTBlob*)->final_start_addr);
       AiwniosAOTBlobDel(cur->ic_data);
     }
diff --git a/Src/AIWNIOS_PrsExp.HC b/Src/AIWNIOS_PrsExp.HC
index 447fb4c..28be815 100644
--- a/Src/AIWNIOS_PrsExp.HC
+++ b/Src/AIWNIOS_PrsExp.HC
@@ -833,6 +833,19 @@ CCodeMisc *CodeMiscHashNew(CCmpCtrl *cc,CHash* h) {
   return cur;
 }
 
+U0 AsmExpressionRelocAdd(CCmpCtrl *cc,CHash *h)
+{
+  CAsmUndefHash *tmpauh;
+
+  if (cc->flags&CCF_ASM_EXPRESSIONS) {
+    tmpauh=CAlloc(sizeof CAsmUndefHash);
+    tmpauh->hash=h;
+    tmpauh->next=cc->asm_undef_hash;
+    cc->asm_undef_hash=tmpauh;
+    cc->flags|=CCF_UNRESOLVED;
+  }
+}
+
 U0 DbgPrint(U8 *fmt,...) {
   U8 *s=StrPrintJoin(NULL,fmt,argc,argv);
   DbgPutS(s);
@@ -996,6 +1009,8 @@ Bool PrsExpression(CCmpCtrl *cc,U8 **ul=NULL,I64 flags=0,CPrsStk *_ps=NULL) { //
   CHashClass *tmpc;
   CMemberLst *mlst;
   CAsmUndefHash *tmpauh;
+  U8 *lookahead;
+  Bool aot_fun_import;
 //u32[0] is type
   //u32[1] is prec
   CRPN *rpn,*rpn2;
@@ -1477,8 +1492,14 @@ found_end_square:
 	  last_consumes_right=0;
 	  Lex(cc);
 	  break;case TK_IDENT:
+	  if(cc->flags&CCF_ASM_EXPRESSIONS&&
+        HashFind(cc->cur_str,cmp.asm_hash,HTT_OPCODE|HTT_ASM_KEYWORD)) {
+	    finish=TRUE;
+	    break;
+	  }
 	  if(finish) break;
 	  last_consumes_right=0;
+          aot_fun_import=FALSE;
 	  if(cc->flags&CCF_ASM_EXPRESSIONS) {
 	    if(cc->local_var_entry&&cc->htc.fun) {
 	      if(flags&PEF_ADDR_OF) {
@@ -1491,12 +1512,29 @@ found_end_square:
 	      tmpex->str=MStrPrint("%s.%s",cc->htc.fun->str,cc->cur_str);
 	      tmpex->type=HTT_EXPORT_SYS_SYM | HTF_UNRESOLVED;
 	      goto ins_label;
-	    } else if(cc->hash_entry&&cc->hash_entry->type&HTT_CLASS) {
+	    } else if(cc->aot&&cc->hash_entry&&
+                  cc->hash_entry->type&HTT_FUN&&
+                  !Bt(&cc->hash_entry(CHashFun*)->flags,Ff_INTERNAL)) {
+              aot_fun_import=TRUE;
+              cc->hash_entry=NULL;
+              goto new;
+            } else if(cc->hash_entry&&cc->hash_entry->type&HTT_CLASS) {
 //MOV RAX, CTask.addr[RBX]
-	      PrsOffsetOf(cc);
-	      break;
+              lookahead=cc->cur_buf_ptr;
+              while(*lookahead&&*lookahead<=' ') lookahead++;
+              if(cc->flags&CCF_USE_LAST_U16&&cc->last_U16=='.'||
+                    *lookahead=='.') {
+	        PrsOffsetOf(cc);
+	        break;
+              }
+              cc->hash_entry=NULL;
+              goto new;
 	    } else if(!cc->hash_entry) {
 new:
+	      if(flags&PEF_ADDR_OF) {
+	        PrsPop(cc->ps);
+	        RPNDel(PrsPop2(cc->ps));
+	      }
 	      tmpex=CAlloc(sizeof(CHashExport));
 	      tmpex->str=StrNew(cc->cur_str);
 	      if (*tmpex->str=='@' && tmpex->str[1]=='@') {
@@ -1512,11 +1550,13 @@ new:
 	      } else {
 	        tmpex->type=HTT_EXPORT_SYS_SYM | HTF_UNRESOLVED;
 ins_label:
-	        tmpc=HashFind(tmpex->str,cc->htc.glbl_hash_table,HTT_EXPORT_SYS_SYM);
-	        if (tmpc)
-		  HashAddAfter(tmpex,tmpc,cc->htc.glbl_hash_table);
-	        else
-		  HashAdd(tmpex,cc->htc.glbl_hash_table);
+                if (!aot_fun_import) {
+	          tmpc=HashFind(tmpex->str,cc->htc.glbl_hash_table,HTT_EXPORT_SYS_SYM);
+	          if (tmpc)
+		    HashAddAfter(tmpex,tmpc,cc->htc.glbl_hash_table);
+	          else
+		    HashAdd(tmpex,cc->htc.glbl_hash_table);
+                }
 	      }
 	      cc->flags|=CCF_UNRESOLVED;
 	      rpn=RPNAddNew(
@@ -1538,6 +1578,7 @@ ins_label:
 		    CodeMiscHashNew(cc,cc->hash_entry),
 		    cmp.internal_types[RT_I64],
 		    );
+	      AsmExpressionRelocAdd(cc,cc->hash_entry);
 	      Lex(cc);
 	      goto fin_ident;
 	    } else if(cc->hash_entry->type&HTT_EXPORT_SYS_SYM) {
@@ -1588,6 +1629,7 @@ ins_label:
 		      CodeMiscHashNew(cc,cc->hash_entry),
 		      cc->hash_entry,
 		      );
+	          AsmExpressionRelocAdd(cc,cc->hash_entry);
 		      #ifdef TARGET_RISCV
 		      rpn->ic_data(CCodeMisc*)->flags|=CMF_SHORT_ADDR;
 		      #endif
@@ -1656,20 +1698,31 @@ ins_label:
 	        goto fin_ident;
 	      }
 	    } else if(cc->hash_entry->type&HTT_GLBL_VAR) {
-	      if(cc->hash_entry(CHashGlblVar*)->data_addr&&!(cc->flags&CCF_AOT_COMPILE)) {
+	      if(cc->hash_entry(CHashGlblVar*)->data_addr&&
+                    !(cc->flags&CCF_AOT_COMPILE)&&
+                    !(cc->aot&&cc->flags&CCF_ASM_EXPRESSIONS)) {
 	        rpn=RPNAddNew(
 		      cc,
 		      IC_IMM_I64,
 		      cc->hash_entry(CHashGlblVar*)->data_addr,
 		      cmp.internal_types[RT_I64],
 		      );
-	      } else if(cc->flags&CCF_AOT_COMPILE) {
+	      } else if(cc->flags&CCF_AOT_COMPILE||
+                    cc->aot&&cc->flags&CCF_ASM_EXPRESSIONS) {
+	        if (cc->flags&CCF_AOT_COMPILE)
+	          tmpex=cc->hash_entry;
+	        else {
+	          tmpex=CAlloc(sizeof CHashExport);
+	          tmpex->str=StrNew(cc->hash_entry->str);
+	          tmpex->type=HTT_EXPORT_SYS_SYM|HTF_UNRESOLVED;
+	        }
 	        rpn=RPNAddNew(
 		      cc,
 		      IC_ADDR_IMPORT,
-		      CodeMiscHashNew(cc,cc->hash_entry),
+		      CodeMiscHashNew(cc,tmpex),
 		      cc->hash_entry(CHashGlblVar*)->var_class,
 		      );
+	        AsmExpressionRelocAdd(cc,tmpex);
 	      } else {
 		  rpn=RPNAddNew(
 		      cc,
@@ -1702,12 +1755,21 @@ ins_label:
 	      rpn->ic_dim=cc->hash_entry(CHashGlblVar*)->dim.next;
 	      Lex(cc);
 	    } else if(cc->hash_entry->type&HTT_EXPORT_SYS_SYM) {
-	        rpn=RPNAddNew(
-		      cc,
-		      IC_IMM_I64,
-		      cc->hash_entry(CHashExport*)->val,
-		      cmp.internal_types[RT_I64],
-		      );
+              if (cc->flags&CCF_AOT_COMPILE &&
+                    !(cc->hash_entry->type&HTF_IMM))
+                rpn=RPNAddNew(
+                      cc,
+                      IC_ADDR_IMPORT,
+                      CodeMiscHashNew(cc,cc->hash_entry),
+                      cmp.internal_types[RT_I64],
+                      );
+              else
+                rpn=RPNAddNew(
+                      cc,
+                      IC_IMM_I64,
+                      cc->hash_entry(CHashExport*)->val,
+                      cmp.internal_types[RT_I64],
+                      );
 	      Lex(cc);
 	    } else if(cc->hash_entry->type&HTT_KEYWORD) {
 	      switch(cc->hash_entry(CHashGeneric *)->user_data0) {
@@ -1802,11 +1864,18 @@ U8 *LexExpression2Bin(CCmpCtrl *cc,I64 *_type=NULL)
 {//Compile cc expression. You call the code.
   U8 *res;
   I64 size;
+  CRPN *rpn;
   CHashFun *old_fun;
   Bool old_trace=LBtr(&cc->flags,CCf_PASS_TRACE_PRESENT);
   COCPush(cc);
   COCInit(cc);
   if (PrsExpression(cc,NULL,PEF_NO_COMMA)) {
+    if (Bt(&cc->opts,OPTf_TEMPLEOS))
+      for (rpn=cc->coc.coc_head.next;rpn!=&cc->coc.coc_head;rpn=rpn->next)
+        if (intermediate_code_table[rpn->type].not_const) {
+          cc->flags|=CCF_NOT_CONST;
+          break;
+        }
     RPNAddNew(cc,IC_RET,0,AssignRawTypeToNode(cc->coc.coc_head.next),FALSE);
     old_fun=cc->htc.fun; //Changed by nroot
     cc->htc.fun=NULL;
diff --git a/Src/AsmResolve.HC b/Src/AsmResolve.HC
index 5b29956..e649413 100644
--- a/Src/AsmResolve.HC
+++ b/Src/AsmResolve.HC
@@ -1,7 +1,130 @@
+U0 AsmUnresolvedAdd(CCmpCtrl *cc,U8 *machine_code,I64 type,I64 rip,I64 rel_rip,
+      CAsmUndefHash *local_asm_undef_hash,CAsmUndefHash *glbl_asm_undef_hash,
+      I64 line_num,Bool U8_avail)
+{
+  CAsmUnresolvedRef *tmpu=MAlloc(sizeof(CAsmUnresolvedRef));
+  tmpu->machine_code=machine_code;
+  tmpu->type=type;
+  tmpu->rip=rip;
+  tmpu->rel_rip=rel_rip;
+  tmpu->aot=cc->aot;
+  tmpu->imm_flag=FALSE;
+  tmpu->line_num=line_num;
+  tmpu->U8_avail=U8_avail;
+  tmpu->str=NULL;
+  if (local_asm_undef_hash) {
+    tmpu->asm_undef_hash=local_asm_undef_hash;
+    tmpu->next=cc->aotc->local_unresolved;
+    cc->aotc->local_unresolved=tmpu;
+  } else {
+    tmpu->asm_undef_hash=glbl_asm_undef_hash;
+    tmpu->next=cc->aotc->glbl_unresolved;
+    cc->aotc->glbl_unresolved=tmpu;
+    if (glbl_asm_undef_hash->hash->type & HTF_IMPORT) {
+      tmpu->str=StrNew(glbl_asm_undef_hash->hash->str);
+      if (glbl_asm_undef_hash->hash->type & HTF_IMM)
+        tmpu->imm_flag=TRUE;
+    }
+  }
+}
+
+U0 AsmResolve(CCmpCtrl *cc,CAsmUnresolvedRef *tmpu,U8 *label,Bool undefined)
+{
+  CAOTImportExport *tmpie;
+  CAOTAbsAddr *tmpabs;
+  CAsmUndefHash *tmpauh;
+  I64 res=Call(tmpu->machine_code);
+  if (undefined) {
+    tmpauh=tmpu->asm_undef_hash;
+    while (tmpauh) {
+      if (tmpauh->hash->type & HTF_UNRESOLVED) {
+        tmpie=CAlloc(sizeof(CAOTImportExport));
+        tmpie->next=tmpauh->hash->ie_lst;
+        tmpauh->hash->ie_lst=tmpie;
+        tmpie->rip=tmpu->rip;
+        tmpie->aot=cc->aot;
+        tmpie->type=tmpu->type;
+      }
+      tmpauh=tmpauh->next;
+    }
+  } else if (!(tmpu->type&IEF_IMM_NOT_REL)) {
+    res-=tmpu->rel_rip;
+    if (tmpu->type==IET_REL_I8 && !(I8_MIN<=res<=I8_MAX) ||
+          tmpu->type==IET_REL_I16 && !(I16_MIN<=res<=I16_MAX)) {
+      PrintErr("Branch out of range at line:%04d %s\n",
+            tmpu->line_num,label);
+      LexExcept(cc);
+    }
+    if (tmpu->U8_avail && tmpu->type>IET_IMM_U8 && -124<=res<=123)
+      PrintWarn("could use I8 displacement at line:%04d %s %s\n",
+            tmpu->line_num,cc->aotc->last_label,label);
+  } else if (tmpu->type==IET_IMM_U32) {
+    for (tmpauh=tmpu->asm_undef_hash;tmpauh;tmpauh=tmpauh->next)
+      if (!(tmpauh->hash->type&HTF_IMM)) {
+        tmpabs=CAlloc(sizeof(CAOTAbsAddr));
+        tmpabs->next=cc->aotc->abss;
+        tmpabs->type=AAT_ADD_U32;
+        tmpabs->rip=tmpu->rip;
+        cc->aotc->abss=tmpabs;
+        break;
+      }
+  }
+  AOTStoreCodeU8At(cc,tmpu->rip,res.u8[0]);
+  if (tmpu->type>=IET_REL_I16) {
+    AOTStoreCodeU8At(cc,tmpu->rip+1,res.u8[1]);
+    if (tmpu->type>=IET_REL_I32) {
+      AOTStoreCodeU8At(cc,tmpu->rip+2,res.u8[2],res.u8[3]);
+      if (tmpu->type>=IET_REL_I64)
+        AOTStoreCodeU8At(cc,tmpu->rip+4,res.u8[4],res.u8[5],
+              res.u8[6],res.u8[7]);
+    }
+  }
+}
+
+U0 AsmExpressionImportsResolve(U8 *name,U8 *module_base,I64 value)
+{
+  CHashTable *table=Fs->hash_table;
+  CHashImport *entry;
+  I64 i,rel_off,resolved;
+
+  for (i=0;i<=table->mask;i++) {
+    entry=table->body[i];
+    while (entry) {
+      if (entry->type&HTT_IMPORT_SYS_SYM &&
+            entry->module_base==module_base &&
+            !StrCmp(entry->str,name)) {
+        rel_off=entry->module_header_entry[1](I32);
+        resolved=value+entry->module_header_entry[5](I32);
+        WriteProtectMemCpy(entry->module_base+rel_off,&resolved,8);
+        entry->type=HTT_INVALID;
+      }
+      entry=entry->next;
+    }
+  }
+}
+
+CHashExport *AOTResolvedAsmExportFind(CCmpCtrl *cc,U8 *name)
+{
+  CHashTable *table;
+  CHash *entry;
+  I64 i;
+
+  table=cc->htc.glbl_hash_table;
+  for (i=0;i<=table->mask;i++)
+    for (entry=table->body[i];entry;entry=entry->next)
+      if (entry->type&HTT_EXPORT_SYS_SYM && entry->type&HTF_EXPORT &&
+              !(entry->type&HTF_UNRESOLVED) &&
+              !StrCmp(entry->str,name))
+          return entry;
+  return NULL;
+}
+
 U0 AOTGlblsResolve(CCmpCtrl *cc,CAOT *tmpaot)
 {
   CAOTCtrl *aotc=cc->aotc;
   CHashFun *tmpf;
+  CHashGlblVar *tmpg;
+  CHashExport *resolved_export,*binding;
   CAsmUnresolvedRef *tmpu=aotc->glbl_unresolved,*tmpu1;
   I64 i,j;
   CAOTImportExport *tmpie,*tmpie1;
@@ -21,10 +144,35 @@ U0 AOTGlblsResolve(CCmpCtrl *cc,CAOT *tmpaot)
 	tmpex=tmpauh->hash;
 	if (tmpex->type & HTT_EXPORT_SYS_SYM&& tmpex->type & HTF_UNRESOLVED &&
 	      !(tmpex->type & HTF_IMPORT) &&
-	      (tmpf=HashFind(tmpex->str,cc->htc.hash_table_lst,HTT_FUN)) &&
+	      (tmpf=HashFind(tmpex->str,cc->htc.glbl_hash_table,HTT_FUN)) &&
 	      !Bt(&tmpf->flags,Cf_EXTERN)) {
 	  tmpex->val=tmpf->exe_addr;
 	  tmpex->type&=~HTF_UNRESOLVED;
+          AsmExpressionImportsResolve(tmpex->str,tmpu->machine_code,
+                tmpf->exe_addr);
+	  label=tmpauh->hash->str;
+        } else if (tmpex->type&HTT_EXPORT_SYS_SYM &&
+              tmpex->type&HTF_UNRESOLVED &&
+              !(tmpex->type&HTF_IMPORT) &&
+              (tmpg=HashFind(tmpex->str,cc->htc.glbl_hash_table,
+                    HTT_GLBL_VAR)) && !(tmpg->flags&GVF_EXTERN)) {
+          tmpex->val=tmpg->data_addr_rip;
+          tmpex->type&=~HTF_UNRESOLVED;
+          AsmExpressionImportsResolve(tmpex->str,tmpu->machine_code,
+                tmpg->data_addr_rip);
+          label=tmpauh->hash->str;
+	} else if (tmpex->type&HTT_EXPORT_SYS_SYM &&
+	      (resolved_export=AOTResolvedAsmExportFind(cc,tmpex->str))) {
+	  tmpex->val=resolved_export->val;
+	  tmpex->type&=~HTF_UNRESOLVED;
+          binding=CAlloc(sizeof CHashExport);
+          binding->type=HTT_EXPORT_SYS_SYM|HTF_RESOLVE;
+          binding->str=StrNew(tmpex->str);
+          binding->val=resolved_export->val;
+          HashAdd(binding,Fs->hash_table);
+          AsmExpressionImportsResolve(tmpex->str,tmpu->machine_code,
+                resolved_export->val);
+          HashRemDel(binding,Fs->hash_table);
 	  label=tmpauh->hash->str;
 	} else {
 	  if (!(tmpex->type & HTF_IMPORT)) {
@@ -48,7 +196,7 @@ U0 AOTGlblsResolve(CCmpCtrl *cc,CAOT *tmpaot)
 	label=tmpauh->hash->str;
       tmpauh=tmpauh->next;
     }
-    //AsmResolve(cc,tmpu,label,undefined); TODO restore ???
+    AsmResolve(cc,tmpu,label,undefined);
     Free(tmpu->machine_code);
     LinkedLstDel(tmpu->asm_undef_hash);
     Free(tmpu);
@@ -77,7 +225,8 @@ U0 AOTGlblsResolve(CCmpCtrl *cc,CAOT *tmpaot)
 	  PrintErr("Undefined sym at %s\n",tmpex->str);
 	  LexExcept(cc);
 	}
-	if (tmpex->type & HTF_RESOLVE) {
+	if (tmpex->type & HTF_RESOLVE &&
+	      tmpex->type & HTT_FUN) {
 	  tmpf=tmpex;
 	  tmpeu=tmpf->ext_lst;
 	  while (tmpeu) {
diff --git a/Src/AsmX64.HC b/Src/AsmX64.HC
index eeeae8a..d4c2b35 100644
--- a/Src/AsmX64.HC
+++ b/Src/AsmX64.HC
@@ -10,7 +10,8 @@ class CAiwniosAOTBlob {
   I64 len;
   U8 *data;
   //private,used when placing ->data into it's final location
-  U8 **final_start_addr;
+  I64 final_start_rip;
+  U8 *final_start_addr;
 };
 U0 AiwniosAOTBlobDel(CAiwniosAOTBlob *b) {
   QueDel(b->asm_patches);
@@ -66,7 +67,7 @@ Bool PrsAsmImm(CCmpCtrl *cc,CX86AddrMode *arg)
   arg->glbl_asm_undef_hash=NULL;
   cc->asm_undef_hash=NULL;
   cc->abs_cnts=0;
-  cc->flags&=~(CCF_UNRESOLVED+CCF_LOCAL);
+  cc->flags&=~(CCF_UNRESOLVED+CCF_LOCAL+CCF_AOT_COMPILE);
   cc->flags|=CCF_ASM_EXPRESSIONS;
   if (!IsLexExpression2Bin(cc,&arg->machine_code))
     LexSkipEol(cc);
@@ -82,7 +83,7 @@ Bool PrsAsmImm(CCmpCtrl *cc,CX86AddrMode *arg)
 	cc->asm_undef_hash=NULL;
       }
     } else {
-      arg->imm=(*arg->machine_code)();
+      arg->imm=Call(arg->machine_code);
       arg->glbl_asm_undef_hash=cc->asm_undef_hash;
       cc->asm_undef_hash=NULL;
       Free(arg->machine_code);
@@ -1071,6 +1072,9 @@ rex|=0x40;
     to[len++]=0x66;
   }
 //Prefix group 4(Address size ovveride)
+  //There is no 16-bit ModRM encoder. TempleOS's 16-bit boot code keeps the
+  //upper halves clear, so address-size override plus the 32-bit form encodes
+  //its [SI]-style operands correctly.
   if (seg_size==16&&indir_mode ||
         seg_size==64&&indir_mode&&indir_mode->short_addr==1)
     to[len++]=0x67;
@@ -1245,9 +1249,10 @@ U0 AddBytesToBuf(U8 **_buf,I64 *len,I64 cnt,U8 *bytes) {
 class CToFill:CQue {
   CHashExport *h;
 };
-U0 AsmResolvePatches(CCmpCtrl *cc,CQue *patches) {
+U0 AsmResolvePatches(CCmpCtrl *cc,CQue *patches,CQue *to_fill) {
   CAsmPatch *patch;
   I64 b,i;
+  CToFill *tmp_to_fill;
   CAsmUndefHash *tab,*cur;
   CHashImport *h;
   CMemberLst *mlst;
@@ -1267,10 +1272,23 @@ U0 AsmResolvePatches(CCmpCtrl *cc,CQue *patches) {
     if(!tab) goto nxt;
     cur=tab;
     while(cur) {
+      for(tmp_to_fill=to_fill->next;tmp_to_fill!=to_fill;
+            tmp_to_fill=tmp_to_fill->next) {
+        if(!StrCmp(tmp_to_fill->h->str,cur->hash->str)) {
+          exp=CAlloc(sizeof CHashExport);
+          exp->type=HTT_EXPORT_SYS_SYM|HTF_RESOLVE;
+          exp->str=StrNew(tmp_to_fill->h->str);
+          exp->val=tmp_to_fill->h->val;
+          HashAdd(exp,Fs->hash_table);
+          SysSymImportsResolve(cur->hash->str,0);
+          HashRemDel(exp,Fs->hash_table);
+          goto resolved;
+        }
+      }
       if(cc->coc.coc_next_misc)
         for(cm=cc->coc.coc_next_misc;cm!=&cc->coc.coc_next_misc;cm=cm->next) {
 	  if(cm->addr2!=INVALID_PTR&&
-	        cm->type==CMT_GOTO_LABEL&&cm->str&&
+	        (cm->type==CMT_GOTO_LABEL||cm->type==CMT_ASM_LABEL)&&cm->str&&
 	        !StrCmp(cm->str,cur->hash->str)) {
 	    exp=CAlloc(sizeof CHashExport);
 	    exp->type=HTT_EXPORT_SYS_SYM|HTF_RESOLVE;
@@ -1304,7 +1322,7 @@ U0 AsmResolvePatches(CCmpCtrl *cc,CQue *patches) {
 resolved:
       cur=cur->next;
     }
-    patch->imm->off=(*patch->imm->machine_code)();
+    patch->imm->off=Call(patch->imm->machine_code);
     Free(patch->imm->machine_code);
     patch->imm->machine_code=NULL;
     LinkedLstDel(tab);
@@ -1312,55 +1330,150 @@ nxt:
     patch=patch->next;
   }
 }
+Bool AsmPatchResolveLabels(CCmpCtrl *cc,CAsmPatch *patch,CQue *to_fill)
+{
+  CAsmUndefHash *tab,*cur;
+  CToFill *label;
+  CMemberLst *member;
+  CHashExport *exp;
+  I64 fun_name_len=0;
+
+  if (cc->htc.fun)
+    fun_name_len=StrLen(cc->htc.fun->str);
+
+  if (tab=patch->imm->local_asm_undef_hash)
+    ;
+  else
+    tab=patch->imm->glbl_asm_undef_hash;
+  if (!tab)
+    return FALSE;
+
+  for (cur=tab;cur;cur=cur->next) {
+    for (label=to_fill->next;label!=to_fill;label=label->next)
+      if (!StrCmp(label->h->str,cur->hash->str))
+        break;
+    if (label==to_fill && (!fun_name_len ||
+          StrNCmp(cur->hash->str,cc->htc.fun->str,fun_name_len) ||
+          cur->hash->str[fun_name_len]!='.' ||
+          !(member=MemberFind(cur->hash->str+fun_name_len+1,cc->htc.fun))))
+      return FALSE;
+    exp=CAlloc(sizeof CHashExport);
+    exp->type=HTT_EXPORT_SYS_SYM|HTF_RESOLVE;
+    exp->str=StrNew(cur->hash->str);
+    if (label!=to_fill)
+      exp->val=label->h->val;
+    else
+      exp->val=-member->offset;
+    HashAdd(exp,Fs->hash_table);
+    AsmExpressionImportsResolve(cur->hash->str,patch->imm->machine_code,
+          exp->val);
+    HashRemDel(exp,Fs->hash_table);
+  }
+
+  patch->imm->off=Call(patch->imm->machine_code);
+  Free(patch->imm->machine_code);
+  patch->imm->machine_code=NULL;
+  LinkedLstDel(tab);
+  patch->imm->local_asm_undef_hash=NULL;
+  patch->imm->glbl_asm_undef_hash=NULL;
+  return TRUE;
+}
+
+U0 AsmPatchStore(CCmpCtrl *cc,U8 *buf,CAsmPatch *patch,I64 value)
+{
+  if (patch->sz==1) buf[patch->rip]=value;
+  else if (patch->sz==2) buf[patch->rip](U16)=value;
+  else if (patch->sz==4) buf[patch->rip](U32)=value;
+  else if (patch->sz==8) buf[patch->rip](U64)=value;
+  else LexExcept(cc,"Invalid assembler relocation size");
+}
+
+I64 AsmPatchRelocationType(CCmpCtrl *cc,CAsmPatch *patch)
+{
+  if (patch->sz==1) {
+      if (patch->is_rel) return IET_REL_I8;
+      else return IET_IMM_U8;
+  } else if (patch->sz==2) {
+      if (patch->is_rel) return IET_REL_I16;
+      else return IET_IMM_U16;
+  } else if (patch->sz==4) {
+      if (patch->is_rel) return IET_REL_I32;
+      else return IET_IMM_U32;
+  } else if (patch->sz==8) {
+      if (patch->is_rel) return IET_REL_I64;
+      else return IET_IMM_I64;
+  }
+  LexExcept(cc,"Invalid AOT assembler relocation size");
+  return 0;
+}
+
 U0 JITFixupAiwniosAOTBlob(CCmpCtrl *cc,CAiwniosAOTBlob *blob,U8 *buf) {
   CToFill *to_fill=blob->to_fill,*tmp_to_fill;
   CAsmPatch *patch,*next_patch;
-  I64 tmp,sz;
+  CAOTAbsAddr *tmpabs;
+  I64 tmp;
   tmp_to_fill=to_fill->next;
   while(tmp_to_fill!=to_fill) {
     tmp_to_fill->h->type&=~HTF_UNRESOLVED;
     tmp_to_fill->h->type|=HTF_RESOLVE;
-    tmp_to_fill->h->val+=buf;
+    if(cc->flags&CCF_AOT_COMPILE)
+      tmp_to_fill->h->val+=blob->final_start_rip;
+    else
+      tmp_to_fill->h->val+=buf;
     tmp_to_fill=tmp_to_fill->next;
   }
-  AsmResolvePatches(cc,blob->asm_patches);
-//If our relocations are not realtive,add the ptr to result
   patch=blob->asm_patches->next;
   if(cc->flags&CCF_AOT_COMPILE) {
-    if(patch!=blob->asm_patches)
-      LexExcept(cc,"AOT assembler not implemented yet for relocs");
+    while(patch!=blob->asm_patches) {
+      next_patch=patch->next;
+      if(patch->imm&&(patch->imm->local_asm_undef_hash||
+            patch->imm->glbl_asm_undef_hash)) {
+        if (AsmPatchResolveLabels(cc,patch,blob->to_fill)) {
+          tmp=patch->imm->off;
+          if (patch->is_rel)
+            tmp-=blob->final_start_rip+patch->rip+patch->sz;
+          AsmPatchStore(cc,buf,patch,tmp);
+          if(!patch->is_rel&&patch->sz==4) {
+            tmpabs=CAlloc(sizeof(CAOTAbsAddr));
+            tmpabs->next=cc->aotc->abss;
+            tmpabs->type=AAT_ADD_U32;
+            tmpabs->rip=blob->final_start_rip+patch->rip;
+            cc->aotc->abss=tmpabs;
+          }
+        } else {
+          AsmUnresolvedAdd(cc,patch->imm->machine_code,
+                AsmPatchRelocationType(cc,patch),
+                blob->final_start_rip+patch->rip,
+                blob->final_start_rip+patch->rip+patch->sz,
+                patch->imm->local_asm_undef_hash,
+                patch->imm->glbl_asm_undef_hash,
+                cc->lex_include_stk->line_num,FALSE);
+          patch->imm->machine_code=NULL;
+          patch->imm->local_asm_undef_hash=NULL;
+          patch->imm->glbl_asm_undef_hash=NULL;
+        }
+      } else if(patch->imm&&(patch->sz==1||patch->sz==2||
+            patch->sz==4||patch->sz==8)) {
+        tmp=patch->imm->off;
+        if(patch->is_rel)
+          tmp-=blob->final_start_rip+patch->rip+patch->sz;
+        AsmPatchStore(cc,buf,patch,tmp);
+        X86AddrModeDel(patch->imm);
+        patch->imm=NULL;
+      }
+      patch=next_patch;
+    }
     return;
   }
+  AsmResolvePatches(cc,blob->asm_patches,blob->to_fill);
+  patch=blob->asm_patches->next;
   while(patch!=blob->asm_patches) {
     next_patch=patch->next;
     if(patch->imm) {
       tmp=patch->imm->off;
-      switch(sz=patch->sz) {
-        case 1:
-	  if(patch->is_rel)
-	    buf[patch->rip]=tmp-(buf+sz+patch->rip);
-	  else
-	    buf[patch->rip]=tmp;
-	  break;
-        case 2:
-	  if(patch->is_rel)
-	    buf[patch->rip](U16)=tmp-(buf+sz+patch->rip);
-	  else
-	    buf[patch->rip](U16)=tmp;
-	  break;
-        case 4:
-	  if(patch->is_rel)
-	    buf[patch->rip](U32)=tmp-(buf+sz+patch->rip);
-	  else
-	    buf[patch->rip](U32)=tmp;
-	  break;
-        case 8:
-	  if(patch->is_rel)
-	    buf[patch->rip](U64)=tmp-(buf+sz+patch->rip);
-	  else
-	    buf[patch->rip](U64)=tmp;
-	  break;
-      }
+      if(patch->is_rel)
+        tmp-=buf+patch->sz+patch->rip;
+      AsmPatchStore(cc,buf,patch,tmp);
       Free(patch->imm);
       patch->imm=NULL;
     }
@@ -1493,7 +1606,25 @@ enter:
       seg_size=64;
       Lex(cc);
     } else if(!StrCmp(cc->hash_entry->str,"IMPORT")) {
-//TODO
+      Lex(cc);
+      while(cc->token&&cc->token!=';') {
+        if(cc->token!=TK_IDENT)
+          LexExcept(cc,"Expected identifier at");
+        exp=CAlloc(sizeof CHashExport);
+        exp->str=cc->cur_str;
+        cc->cur_str=NULL;
+        exp->type=HTT_EXPORT_SYS_SYM|HTF_UNRESOLVED;
+        HashSrcFileSet(cc,exp);
+        HashAdd(exp,cc->htc.glbl_hash_table);
+        exp->type|=HTF_IMPORT;
+        if(Lex(cc)==',')
+          Lex(cc);
+        else if(cc->token!=';')
+          LexExcept(cc,"Expected ',' or ';' at");
+      }
+      if(cc->token!=';')
+        LexExcept(cc,"Missing ';' at");
+      Lex(cc);
     } else if(!StrCmp(cc->hash_entry->str,"BINFILE")) {
       if(Lex(cc)!=TK_STR)
         LexExcept(cc,"Expected a filename.");
diff --git a/Src/CExts.HC b/Src/CExts.HC
index e29ec74..0e079a0 100644
--- a/Src/CExts.HC
+++ b/Src/CExts.HC
@@ -40,4 +40,5 @@ extern I64 PrsUnaryTerm(
 extern U0 PrsVarInit2(CCmpCtrl *cc,U8 **_dst,CHashClass *tmpc,
 	CArrayDim *tmpad,U8 *data_addr_rip,U8 **_base,Bool data_heap,I64 pass,CHash * to);
 extern U0 FillInShortCalls(CCmpCtrl *cc);
+extern CHash *AOTResolvedHashFind(CCmpCtrl *cc,U8 *name);
 extern CHashClass *AssignRawTypeToNode(CRPN *r);
diff --git a/Src/CMain.HC b/Src/CMain.HC
index 065d87b..6766b53 100644
--- a/Src/CMain.HC
+++ b/Src/CMain.HC
@@ -4,12 +4,14 @@ U8 *LexStmt2Bin(CCmpCtrl *cc,I64 *_type,I64 cmp_flags=0)
   CCodeCtrl *tmpcbh;
   CRPN *rpn;
   LBtr(&cc->flags,CCf_PASS_TRACE_PRESENT);
+  cc->flags&=~CCF_RAW_AOT_ASM;
   if (cc->aot_depth==2)
     COCPush(cc);
   COCInit(cc);
   if (!PrsStmt(cc,,,cmp_flags)) {
     if (cc->coc.coc_head.next!=&cc->coc.coc_head) {
       rpn=cc->coc.coc_head.next;
+      if (!(cc->flags&CCF_RAW_AOT_ASM))
       switch(rpn->type) {
         case IC_TO_I64:
         case IC_TO_F64:
@@ -75,7 +77,7 @@ U8 *LexStmt2Bin(CCmpCtrl *cc,I64 *_type,I64 cmp_flags=0)
 	  RPNAddNew(cc,IC_IMM_I64,0,cmp.internal_types[RT_I64],FALSE);
 	  RPNAddNew(cc,IC_RET,0,cmp.internal_types[RT_I64],FALSE);
       }
-      if(cc->flags&CCF_AOT_COMPILE)
+      if(cc->flags&CCF_AOT_COMPILE && !(cc->flags&CCF_RAW_AOT_ASM))
 //ARM likes 4byte align for instrcutions
         while(cc->aotc->rip%4)
 	  AOTStoreCodeU8(cc,0);
@@ -86,9 +88,14 @@ U8 *LexStmt2Bin(CCmpCtrl *cc,I64 *_type,I64 cmp_flags=0)
       if (res) {
         if (cc->flags&CCF_AOT_COMPILE) {
 	  j=cc->aotc->rip;
-	  k=(size+7)>>3;
-	  for (i=0;i<k;i++)
-	    AOTStoreCodeU64(cc,res[i]);
+          if (cc->flags&CCF_RAW_AOT_ASM)
+            for (i=0;i<size;i++)
+              AOTStoreCodeU8(cc,res(U8*)[i]);
+          else {
+            k=(size+7)>>3;
+            for (i=0;i<k;i++)
+              AOTStoreCodeU64(cc,res[i]);
+          }
 	  Free(res);
 	  res=j;
         }
@@ -228,11 +235,12 @@ CAOT *CmpJoin(CCmpCtrl *cc,I64 cmp_flags,U8 *map_name=NULL,U8 mapfile_drv_let=0)
 }
 
 CAOT *CmpBuf(U8 *buf,U8 *map_name=NULL,
-  I64 *error_cnt=NULL, I64 *warning_cnt=NULL,U8 mapfile_drv_let=0)
+  I64 *error_cnt=NULL, I64 *warning_cnt=NULL,U8 mapfile_drv_let=0,I64 opts=0)
 {
   CCmpCtrl *cc;
   CAOT *res=NULL;
   cc=CmpCtrlNew(buf,CCF_DONT_FREE_BUF);
+  cc->opts=opts;
   cc->flags|=CCF_AOT_COMPILE;
   #ifdef TARGET_AARCH64
   //For MacOS RWX protections. 
@@ -254,12 +262,12 @@ CAOT *CmpBuf(U8 *buf,U8 *map_name=NULL,
   return res;
 }
 
-I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0)
+I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0,I64 opts=0)
 {//AOT Compile HC or PRJ file a and output BIN file. Returns err_cnt.
-  U8 *ptr,*fbuf=NULL,*fbuf2=NULL,*fbuf3=NULL,
+  U8 *ptr,*new_buf,*fbuf=NULL,*fbuf2=NULL,*fbuf3=NULL,
 	*patch_table=MAlloc(0x800000);
   CAOT *tmpaot;
-  I64 i,cnt,size=0,error_cnt=0,warning_cnt=0,aot_U8s=0;
+  I64 i,cnt,size=0,data_rip,error_cnt=0,warning_cnt=0,aot_U8s=0;
   CBinFile *bfh;
   CAOTImportExport *tmpie,*tmpie1;
   CAOTAbsAddr *tmpa,*tmpa1;
@@ -271,8 +279,52 @@ I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0)
   if (map_name)
     fbuf3=ExtDft(map_name,"MAP.Z");
 
-  if (tmpaot=CmpBuf(fbuf2,fbuf3,&error_cnt,&warning_cnt,mapfile_drv_let)) {
+  if (tmpaot=CmpBuf(fbuf2,fbuf3,&error_cnt,&warning_cnt,mapfile_drv_let,opts)) {
     aot_U8s=tmpaot->aot_U8s;
+    if (Bt(&opts,OPTf_TEMPLEOS) && tmpaot->heap_glbls) {
+      size=aot_U8s;
+      for (tmphg=tmpaot->heap_glbls;tmphg;tmphg=tmphg->next)
+        size=((size+7)&-8)+tmphg->size;
+      new_buf=MAlloc(size);
+      MemCpy(new_buf,tmpaot->buf,aot_U8s);
+      MemSet(new_buf+aot_U8s,0,size-aot_U8s);
+      Free(tmpaot->buf);
+      tmpaot->buf=new_buf;
+      tmphg=tmpaot->heap_glbls;
+      while (tmphg) {
+        tmphg1=tmphg->next;
+        data_rip=(aot_U8s+7)&-8;
+        if (tmphg->data)
+          MemCpy(tmpaot->buf+data_rip,tmphg->data,tmphg->size);
+        if (tmphg->str) {
+          tmpie=CAlloc(sizeof(CAOTImportExport));
+          tmpie->type=IET_REL32_EXPORT;
+          tmpie->rip=data_rip;
+          tmpie->str=tmphg->str;
+          QueIns(tmpie,tmpaot->last_ie);
+          tmphg->str=NULL;
+        }
+        tmphgr=tmphg->references;
+        while (tmphgr) {
+          tmphgr1=tmphgr->next;
+          *(tmpaot->buf+tmphgr->rip)(I64 *)+=data_rip+tmphgr->offset;
+          tmpa=CAlloc(sizeof(CAOTAbsAddr));
+          tmpa->type=AAT_ADD_U64;
+          tmpa->rip=tmphgr->rip;
+          tmpa->next=tmpaot->abss;
+          tmpaot->abss=tmpa;
+          Free(tmphgr);
+          tmphgr=tmphgr1;
+        }
+        aot_U8s=data_rip+tmphg->size;
+        Free(tmphg->str);
+        Free(tmphg->data);
+        Free(tmphg);
+        tmphg=tmphg1;
+      }
+      tmpaot->heap_glbls=NULL;
+      tmpaot->aot_U8s=aot_U8s;
+    }
     ptr=patch_table;
 //See $LK,"Load",A="MN:Load"$()
     cnt=0;
@@ -337,7 +389,7 @@ I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0)
     tmpie=tmpaot->next_ie;
     while (tmpie!=&tmpaot->next_ie) {
       tmpie1=tmpie->next;
-      if (!tmpie->type || IET_REL32_EXPORT<=tmpie->type<=IET_IMM64_EXPORT) {
+      if (IET_REL32_EXPORT<=tmpie->type<=IET_IMM64_EXPORT) {
 	QueRem(tmpie);
 	*ptr++=tmpie->type;
 	*ptr(U32 *)++=tmpie->rip;
@@ -359,13 +411,17 @@ I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0)
     while (tmpie!=&tmpaot->next_ie) {
       tmpie1=tmpie->next;
       QueRem(tmpie);
+      if (!tmpie->type)
+        goto skip_import;
       *ptr++=tmpie->type;
       if (tmpie->aot)
 	tmpie->rip+=tmpie->aot->rip2;
       *ptr(U32 *)++=tmpie->rip;
-      switch(tmpie->type) {
-      case IET_REL_I0 ... IET_REL_RISCV:
-		*ptr(I32 *)++=tmpie->offset;
+      if (!Bt(&opts,OPTf_TEMPLEOS)) {
+        switch(tmpie->type) {
+        case IET_REL_I0 ... IET_REL_RISCV:
+          *ptr(I32 *)++=tmpie->offset;
+        }
       }
       if (tmpie->str) {
 	i=StrLen(tmpie->str);
@@ -374,6 +430,7 @@ I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0)
 	ptr+=i+1;
       } else
 	*ptr++=0;
+skip_import:
       Free(tmpie->src_link);
       Free(tmpie);
       tmpie=tmpie1;
@@ -405,6 +462,8 @@ I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0)
 #ifdef TARGET_RISCV
     bfh->bin_signature='RV64';
 #endif
+    if (Bt(&opts,OPTf_TEMPLEOS))
+      bfh->bin_signature=BIN_SIGNATURE_VAL;
     MemCpy(bfh(U8 *)+sizeof(CBinFile),tmpaot->buf,aot_U8s);
     MemCpy(bfh(U8 *)+sizeof(CBinFile)+aot_U8s,patch_table,
 	  size-aot_U8s-sizeof(CBinFile));
@@ -551,17 +610,47 @@ I64 StreamExePrint(U8 *fmt,...)
   return res;
 }
 
+CHash *AOTResolvedHashFind(CCmpCtrl *cc,U8 *name)
+{
+  CHash *h;
+  CHashTable *table=cc->htc.glbl_hash_table;
+  I64 i;
+  for (i=0;i<=table->mask;i++)
+    for (h=table->body[i];h;h=h->next)
+      if (h->type&HTT_FUN && !StrCmp(h->str,name) &&
+            h->type&HTF_RESOLVE &&
+            0<=h(CHashFun*)->exe_addr<cc->aotc->rip)
+        return h;
+  for (i=0;i<=table->mask;i++)
+    for (h=table->body[i];h;h=h->next)
+      if (h->type&HTT_GLBL_VAR && !StrCmp(h->str,name) &&
+            !(h(CHashGlblVar*)->flags&GVF_EXTERN))
+        return h;
+  for (i=0;i<=table->mask;i++)
+    for (h=table->body[i];h;h=h->next)
+      if (h->type&HTT_EXPORT_SYS_SYM && !StrCmp(h->str,name) &&
+            h->type&HTF_RESOLVE)
+        return h;
+  return NULL;
+}
+
 //Added by nroot
 U0 FillInShortCalls(CCmpCtrl *cc) {
   CAOTImportExport *tmpaotie,*tmpaotie1;
+  CAOTAbsAddr *tmpabs;
   CHashFun *fun;
+  Bool value_resolved;
   I64 i,ptr,low12;
   for(tmpaotie=cc->aot->next_ie;tmpaotie!=&cc->aot->next_ie;tmpaotie=tmpaotie1) {
     tmpaotie1=tmpaotie->next;
     if(tmpaotie->short_jmp) {		  
-      fun=HashSingleTableFind(tmpaotie->str,cc->htc.glbl_hash_table,HTT_FUN|HTT_GLBL_VAR);
+      if (Bt(&cc->opts,OPTf_TEMPLEOS))
+        fun=AOTResolvedHashFind(cc,tmpaotie->str);
+      else
+        fun=HashSingleTableFind(tmpaotie->str,cc->htc.glbl_hash_table,
+              HTT_FUN|HTT_GLBL_VAR);
       if(fun&&fun->type&HTT_FUN) {
-		if(fun->type&HTF_RESOLVE) {
+		if (fun->type&HTF_RESOLVE) {
 		  #ifdef TARGET_RISCV
 		  ptr=fun->exe_addr+tmpaotie->offset;
 riscv_patch:
@@ -615,8 +704,51 @@ riscv_patch:
 			Free(tmpaotie->str);
 			Free(tmpaotie);
 skip:;
+      } else if (fun && fun->type&HTT_EXPORT_SYS_SYM &&
+            fun->type&HTF_RESOLVE) {
+#ifdef TARGET_X86
+        i=fun(CHashExport*)->val-(tmpaotie->rip+4)+tmpaotie->offset;
+        AOTStoreCodeU8At(cc,tmpaotie->rip,
+              i.u8[0],i.u8[1],i.u8[2],i.u8[3]);
+        QueRem(tmpaotie);
+        Free(tmpaotie->str);
+        Free(tmpaotie);
+#else
+        PrintErr("AOT export relocation is not implemented for this target\n");
+        throw('Compiler');
+#endif
+      }
+    } else if (Bt(&cc->opts,OPTf_TEMPLEOS) &&
+          tmpaotie->type==IET_IMM_I64 && tmpaotie->str) {
+      value_resolved=TRUE;
+      fun=AOTResolvedHashFind(cc,tmpaotie->str);
+      if (fun && fun->type&HTT_FUN && fun->type&HTF_RESOLVE)
+        i=fun->exe_addr+tmpaotie->offset;
+      else if (fun && fun->type&HTT_GLBL_VAR && fun->type&HTF_EXPORT &&
+            !(fun(CHashGlblVar*)->flags&(GVF_EXTERN|GVF_DATA_HEAP)))
+        i=fun(CHashGlblVar*)->data_addr_rip+tmpaotie->offset;
+      else if (fun && fun->type&HTT_EXPORT_SYS_SYM &&
+            fun->type&HTF_RESOLVE)
+        i=fun(CHashExport*)->val+tmpaotie->offset;
+      else
+        value_resolved=FALSE;
+      if (value_resolved) {
+        ptr=tmpaotie->rip;
+        if (tmpaotie->aot)
+          ptr+=tmpaotie->aot->rip2;
+        AOTStoreCodeU8At(cc,ptr,
+              i,i>>8,i>>16,i>>24);
+        AOTStoreCodeU8At(cc,ptr+4,
+              i>>32,i>>40,i>>48,i>>56);
+        tmpabs=CAlloc(sizeof(CAOTAbsAddr));
+        tmpabs->next=cc->aotc->abss;
+        tmpabs->type=AAT_ADD_U64;
+        tmpabs->rip=ptr;
+        cc->aotc->abss=tmpabs;
+        QueRem(tmpaotie);
+        Free(tmpaotie->str);
+        Free(tmpaotie);
       }
     }
   }
 }
-
diff --git a/Src/KLoadARM.HC b/Src/KLoadARM.HC
index 31409d3..075257b 100644
--- a/Src/KLoadARM.HC
+++ b/Src/KLoadARM.HC
@@ -198,6 +198,27 @@ U0 LoadPass1(U8 *src,U8 *module_base,I64 ld_flags)
   }
 }
 
+Bool BinSignatureValid(U32 signature)
+{
+  if (signature==BIN_SIGNATURE_VAL)
+    return TRUE;
+#ifdef TARGET_BYTECODE
+  return signature=='BYTC';
+#endif
+#ifdef TARGET_X86
+  if (HOST_ABI=='OpenBSD')
+    return signature=='OX86';
+  return signature=='X86';
+#endif
+#ifdef TARGET_AARCH64
+  return signature=='ARM';
+#endif
+#ifdef TARGET_RISCV
+  return signature=='RV64';
+#endif
+  return FALSE;
+}
+
 U0 LoadPass2(U8 *src,U8 *module_base,I64)
 {
   U8 *st_ptr;
@@ -250,7 +271,7 @@ CBinFile *Load(U8 *filename,I64 ld_flags=0,CBinFile *bfh_addr=INVALID_PTR)
 
   //See $$LK,"Patch Table Generation",A="FF:::/Compiler/CMain.HC,IET_ABS_ADDR"$$
   module_align=1<<bfh->module_align_bits;
-  if (!module_align || bfh->bin_signature!=BIN_SIGNATURE_VAL) {
+  if (!module_align || !BinSignatureValid(bfh->bin_signature)) {
     Free(bfh);
     Free(fbuf);
     throw('BinModul');
diff --git a/Src/KernelA.HH b/Src/KernelA.HH
index 2b0c63f..8260430 100644
--- a/Src/KernelA.HH
+++ b/Src/KernelA.HH
@@ -1530,6 +1530,8 @@ public class CAutoCompleteDictGlbls
 #define OPTf_KEEP_RELOCATIONS 0x26 //Added by nroot,will keep the CHashImport's
 #define OPTf_BOUNDS_CHECK 0x27 //Added by nroot,Bounds checker
 #define OPTf_BYTE_CODE 0x28 //Added by nroot
+#define OPTf_TEMPLEOS 0x2A //TempleOS x86-64 ABI and module format
+#define OPTf_RAW_AOT 0x29
 
 #define OPTF_ECHO		(1<<OPTf_ECHO)
 
@@ -2171,6 +2173,7 @@ class CLexHashTableContext
 #define CCF_PAREN		0x2000000000
 #define CCf_PAREN		37
 #define CCF_CLASS_DOL_OFFSET	0x4000000000
+#define CCF_RAW_AOT_ASM		0x10000000000
 #define CCF_DONT_MAKE_RES	0x8000000000
 
 public class CCmpCtrl
@@ -4138,7 +4141,7 @@ extern U64 ToUpper(U64);
 extern U8 *FileRead(U8 *,I64 *,I64 *attr=NULL);
 extern U0 FileWrite(U8 *filename,U8 *fbuf,I64 size,CDate cdt=0,I64 attr=0);
 
-extern I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0);
+extern I64 Cmp(U8 *filename,U8 *map_name=NULL,U8 *out_name=NULL,U8 mapfile_drv_let=0,I64 opts=0);
 
 //Forward declarations
 extern Bool DirMk(U8 *filename,I64 entry_cnt=0);
@@ -4969,7 +4972,7 @@ extern U0 CtrlAltCBSet(U8 ch,U0 (*fp_hndlr)(I64 sc),
 	U8 *no_shift_desc=NULL,U8 *shift_desc=NULL,Bool in_irq=FALSE);
 extern U0 EdLineDown(CDoc *doc,I64 sc=I64_MIN);
 extern CAOT *CmpBuf(U8 *buf,U8 *map_name=NULL,
-  I64 *error_cnt=NULL, I64 *warning_cnt=NULL,U8 mapfile_drv_let=0);
+  I64 *error_cnt=NULL, I64 *warning_cnt=NULL,U8 mapfile_drv_let=0,I64 opts=0);
 extern I64 PressAKey();
 extern U8 *DocLineRead(U8 *filename,I64 line,CTask *mem_task=NULL);
 extern U8 *DocLineWrite(U8 *filename,I64 line,U8 *st);
@@ -6916,12 +6919,16 @@ class CAiwniosAOTBlob {
   I64 len;
   U8 *data;
   //private,used when placing ->data into it's final location
-  U8 **final_start_addr;
+  I64 final_start_rip;
+  U8 *final_start_addr;
 };
 extern U0 AiwniosAOTBlobDel(CAiwniosAOTBlob *b);
+extern U0 AsmUnresolvedAdd(CCmpCtrl *cc,U8 *machine_code,I64 type,I64 rip,
+      I64 rel_rip,CAsmUndefHash *local_asm_undef_hash,
+      CAsmUndefHash *glbl_asm_undef_hash,I64 line_num,Bool U8_avail);
 extern U0 JITFixupAiwniosAOTBlob(CCmpCtrl *cc,CAiwniosAOTBlob *blob,U8 *buf);
 extern U0 AsmResolvePatches(CCmpCtrl *cc,CQue *patches);
-extern I64 X86EncodeOpcode(I64 rip,U8 *opc,U8 **modes,U8 *to,CQue *asm_patches);
+extern I64 X86EncodeOpcode(I64 rip,I64 seg_size,U8 *opc,U8 **modes,U8 *to,CQue *asm_patches);
 #ifdef TARGET_X86
 extern U8 SSE_ShuffleByte(I64 a,I64 b,I64 c,I64 d);
 #endif
diff --git a/Src/PrsStmt.HC b/Src/PrsStmt.HC
index 7d8c9f7..b863526 100644
--- a/Src/PrsStmt.HC
+++ b/Src/PrsStmt.HC
@@ -453,6 +453,9 @@ U0 PrsGlblVarLst(CCmpCtrl *cc,I64 saved_mode,CHashClass *saved_tmpc,
 	PrsGlblInit(cc,tmpg,2);
 	if (cc->flags&CCF_AOT_COMPILE) {
 	  if (!Bt(&cc->opts,OPTf_GLBLS_ON_DATA_HEAP)) {
+	    if (Bt(&cc->opts,OPTf_TEMPLEOS))
+	      for (k=0;k<tmpg->size;k++)
+		AOTStoreCodeU8At(cc,tmpg->data_addr_rip+k,tmpg->data_addr[k]);
 	  } else
 //Nroot is modifing Load to include the ->data_addr in the IET_DATA_HEAP and in loader.c
 	    tmpg->heap_glbl->data=MAllocIdent(tmpg->data_addr);
@@ -1041,13 +1044,15 @@ CmpFixUpJITAsm(cc,tmpaot);
 		    if (tmpaot=PrsAsmBlk(cc)) {
 		      //If we are not in a function,we will jump past the assembly block
 		      // to avoid running it
-		      if(!cc->htc.fun) {
+		      if (!cc->htc.fun && Bt(&cc->opts,OPTf_RAW_AOT))
+			cc->flags|=CCF_RAW_AOT_ASM;
+		      else if(!cc->htc.fun) {
 		        g_lb=COCMiscNew(cc,CMT_LABEL);
 		        RPNAddNew(cc,IC_JMP,g_lb,NULL,FALSE);
 		      }
 		      RPNAddNew(cc,IC_ASM,tmpaot,NULL,FALSE);
 		      cc->flags|=CCF_HAS_MISC_DATA;
-		      if(!cc->htc.fun)
+		      if(!cc->htc.fun && !Bt(&cc->opts,OPTf_RAW_AOT))
 		        RPNAddNew(cc,IC_LABEL,g_lb,NULL,FALSE);
 		    }
 #endif
@@ -1055,13 +1060,15 @@ CmpFixUpJITAsm(cc,tmpaot);
 		    if (tmpaot=PrsAsmBlk(cc)) {
 		      //If we are not in a function,we will jump past the assembly block
 		      // to avoid running it
-		      if(!cc->htc.fun) {
+		      if (!cc->htc.fun && Bt(&cc->opts,OPTf_RAW_AOT))
+			cc->flags|=CCF_RAW_AOT_ASM;
+		      else if(!cc->htc.fun) {
 		        g_lb=COCMiscNew(cc,CMT_LABEL);
 		        RPNAddNew(cc,IC_JMP,g_lb,NULL,FALSE);
 		      }
 		      RPNAddNew(cc,IC_ASM,tmpaot,NULL,FALSE);
 		      cc->flags|=CCF_HAS_MISC_DATA;
-		      if(!cc->htc.fun)
+		      if(!cc->htc.fun && !Bt(&cc->opts,OPTf_RAW_AOT))
 		        RPNAddNew(cc,IC_LABEL,g_lb,NULL,FALSE);
 		    }
 #endif
@@ -1069,13 +1076,15 @@ CmpFixUpJITAsm(cc,tmpaot);
 		    if (tmpaot=PrsAsmBlk(cc)) {
 		      //If we are not in a function,we will jump past the assembly block
 		      // to avoid running it
-		      if(!cc->htc.fun) {
+		      if (!cc->htc.fun && Bt(&cc->opts,OPTf_RAW_AOT))
+			cc->flags|=CCF_RAW_AOT_ASM;
+		      else if(!cc->htc.fun) {
 		        g_lb=COCMiscNew(cc,CMT_LABEL);
 		        RPNAddNew(cc,IC_JMP,g_lb,NULL,FALSE);
 		      }
 		      RPNAddNew(cc,IC_ASM,tmpaot,NULL,FALSE);
 		      cc->flags|=CCF_HAS_MISC_DATA;
-		      if(!cc->htc.fun)
+		      if(!cc->htc.fun && !Bt(&cc->opts,OPTf_RAW_AOT))
 		        RPNAddNew(cc,IC_LABEL,g_lb,NULL,FALSE);
 		    }
 #endif
diff --git a/Src/PrsVar.HC b/Src/PrsVar.HC
index 7893867..ea72454 100644
--- a/Src/PrsVar.HC
+++ b/Src/PrsVar.HC
@@ -43,11 +43,9 @@ U0 PrsVarInit(CCmpCtrl *cc,U8 **_dst,CHashClass *tmpc,CArrayDim *tmpad,
       cc->flags=CCF_NO_ABSS | cc->flags &
 	    ~(CCF_AOT_COMPILE|CCF_HAS_MISC_DATA|CCF_NOT_CONST);
       machine_code=LexExpression2Bin(cc,&type);
-      if (old_flags&CCF_AOT_COMPILE
-	    /*cc->flags&CCF_NOT_CONST && */ //Const-folding is done in the C end
-	//Nroot forces all AOT globals to go on the heap for Apple hardedned runtime
-	    /*!Bt(&cc->opts,OPTf_GLBLS_ON_DATA_HEAP)*/
-	) {
+      if (old_flags&CCF_AOT_COMPILE &&
+            (!Bt(&cc->opts,OPTf_TEMPLEOS) ||
+            cc->flags&CCF_NOT_CONST)) {
 	cc->flags=cc->flags&~CCF_NO_ABSS|CCF_AOT_COMPILE;
 	Free(machine_code);
 	if (pass==2) {
@@ -145,6 +143,10 @@ U0 PrsVarInit2(CCmpCtrl *cc,U8 **_dst,CHashClass *tmpc,
       MemCpy(*_base,st,i);
 	*_dst=*_base+i;
       } else {
+	if (cc->flags&CCF_AOT_COMPILE && pass==2)
+	  for (j=0;j<tmpad1->cnt;j++)
+	    AOTStoreCodeU8At(cc,data_addr_rip+*_dst-*_base+j,
+		  st[j]);
 	MemCpy(*_dst,st,tmpad1->cnt);
 	*_dst+=tmpad1->cnt;
       }
@@ -598,8 +600,6 @@ pvl_restart3:
 	    tmpm->use_cnt=0;
 	    cc->flags|=CCF_DONT_MAKE_RES;
 	  }
-	  if (cc->flags&CCF_AOT_COMPILE)
-	    Free(tmpm->static_data);
 	  break;
 	case PRS1B_LOCAL_VAR:
 	  if (mode&PRSF_UNION) {
diff --git a/c/aiwn_lexparser.h b/c/aiwn_lexparser.h
index 75013d2..0854a0f 100644
--- a/c/aiwn_lexparser.h
+++ b/c/aiwn_lexparser.h
@@ -185,6 +185,7 @@ typedef struct CCmpCtrl {
 #define CCF_STRINGS_ON_HEAP          0x2
 #define CCF_AOT_COMPILE              0x4
 #define CCF_ICMOV_NO_USE_RAX         0x8
+#define CCF_RAW_AOT_ASM              0x10
   int64_t flags;
   CHashFun *cur_fun;
   CCodeCtrl *code_ctrl;
@@ -764,6 +765,7 @@ extern void *GenFFIBindingNaked(void *fptr, int64_t arity);
 extern void PrsBindCSymbolNaked(char *name, void *ptr, int64_t arity);
 extern void PrsBindCSymbol(char *name, void *ptr, int64_t arity);
 void __HC_CmpCtrl_SetAOT(CCmpCtrl *cc);
+void __HC_CmpCtrl_SetRawAOTAsm(CCmpCtrl *cc);
 
 CRPN *__HC_ICAdd_GetVargsPtr(CCodeCtrl *cc);
 
diff --git a/c/arm_backend.c b/c/arm_backend.c
index 50dc82d..11ee9fc 100644
--- a/c/arm_backend.c
+++ b/c/arm_backend.c
@@ -5171,7 +5171,8 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
           obin); // For OpenBSD sexy mapping(see mem.c or ask nroot)
       code_off = 0;
     }
-    code_off = FuncProlog(cctrl, bin, code_off);
+    if (!(cctrl->flags & CCF_RAW_AOT_ASM))
+      code_off = FuncProlog(cctrl, bin, code_off);
     for (cnt = 0; cnt < cnt2; cnt++) {
     enter:
       cctrl->backend_user_data1 = 0;
@@ -5191,7 +5192,8 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
       }
     }
     cctrl->epilog_label->addr = code_off + bin;
-    code_off = FuncEpilog(cctrl, bin, code_off);
+    if (!(cctrl->flags & CCF_RAW_AOT_ASM))
+      code_off = FuncEpilog(cctrl, bin, code_off);
     for (misc = cctrl->code_ctrl->code_misc->next;
          misc != cctrl->code_ctrl->code_misc; misc = misc->base.next) {
       switch (misc->type) {
@@ -5301,7 +5303,7 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
       }
     }
     if (run) {
-      if (code_off % 8) // Align to 8
+      if (!(cctrl->flags & CCF_RAW_AOT_ASM) && code_off % 8) // Align to 8
         code_off += 8 - code_off % 8;
       cctrl->statics_label->addr = code_off + bin;
       for (r = cctrl->code_ctrl->ir_code->next; r != cctrl->code_ctrl->ir_code;
diff --git a/c/main.c b/c/main.c
index 86a6841..a1814b7 100644
--- a/c/main.c
+++ b/c/main.c
@@ -918,6 +918,11 @@ static int64_t STK___HC_CmpCtrl_SetAOT(int64_t *stk) {
   __HC_CmpCtrl_SetAOT((CCmpCtrl *)stk[0]);
 }
 
+static int64_t STK___HC_CmpCtrl_SetRawAOTAsm(int64_t *stk) {
+  __HC_CmpCtrl_SetRawAOTAsm((CCmpCtrl *)stk[0]);
+  return 0;
+}
+
 static int64_t STK__HC_ICAdd_RawBytes(int64_t *stk) {
   return (int64_t)__HC_ICAdd_RawBytes((CCmpCtrl *)stk[0], (char *)stk[1],
                                       stk[2]);
@@ -1729,6 +1734,7 @@ static void BootAiwnios(char *bootstrap_text) {
     PrsAddSymbol("__HC_ICAdd_GetVargsPtr", STK___HC_ICAdd_GetVargsPtr, 1);
     PrsAddSymbol("IsValidPtr", STK_IsValidPtr, 1);
     PrsAddSymbol("__HC_CmpCtrl_SetAOT", STK___HC_CmpCtrl_SetAOT, 1);
+    PrsAddSymbol("__HC_CmpCtrl_SetRawAOTAsm", STK___HC_CmpCtrl_SetRawAOTAsm, 1);
     PrsAddSymbol("__HC_ICAdd_Typecast", STK___HC_ICAdd_Typecast, 3);
     PrsAddSymbol("__HC_ICAdd_SubCall", STK___HC_ICAdd_SubCall, 2);
     PrsAddSymbol("__HC_ICAdd_SubProlog", STK___HC_ICAdd_SubProlog, 1);
diff --git a/c/parser.c b/c/parser.c
index 9acea7e..1454873 100644
--- a/c/parser.c
+++ b/c/parser.c
@@ -5082,6 +5082,10 @@ CRPN *__HC_ICAdd_StaticRef(CCodeCtrl *cc, int64_t off, int64_t rt,
 void __HC_CmpCtrl_SetAOT(CCmpCtrl *cc) {
   cc->flags |= CCF_AOT_COMPILE;
 }
+
+void __HC_CmpCtrl_SetRawAOTAsm(CCmpCtrl *cc) {
+  cc->flags |= CCF_RAW_AOT_ASM;
+}
 HC_IC_BINDING(HC_ICAdd_ToBool, IC_TO_BOOL);
 HC_IC_BINDING(HC_ICAdd_BT, IC_BT);
 HC_IC_BINDING(HC_ICAdd_BTC, IC_BTC);
diff --git a/c/riscv_backend.c b/c/riscv_backend.c
index 2387999..fed1468 100644
--- a/c/riscv_backend.c
+++ b/c/riscv_backend.c
@@ -4721,7 +4721,8 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
       bin = A_MALLOC(1024 + code_off, heap ? heap : Fs->code_heap);
       code_off = 0;
     }
-    code_off = FuncProlog(cctrl, bin, code_off);
+    if (!(cctrl->flags & CCF_RAW_AOT_ASM))
+      code_off = FuncProlog(cctrl, bin, code_off);
     for (cnt = 0; cnt < cnt2; cnt++) {
     enter:
       r = forwards[cnt];
@@ -4749,7 +4750,8 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
     }
     cctrl->epilog_label->addr = bin + code_off;
     cctrl->epilog_label->code_off = code_off;
-    code_off = FuncEpilog(cctrl, bin, code_off);
+    if (!(cctrl->flags & CCF_RAW_AOT_ASM))
+      code_off = FuncEpilog(cctrl, bin, code_off);
     for (misc = cctrl->code_ctrl->code_misc->next;
          misc != cctrl->code_ctrl->code_misc; misc = misc->base.next) {
       int64_t old_code_off = code_off;
@@ -4885,7 +4887,7 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
         goto fill_in_refs;
       }
     }
-    if (code_off % 8) // Align to 8
+    if (!(cctrl->flags & CCF_RAW_AOT_ASM) && code_off % 8) // Align to 8
       code_off += 8 - code_off % 8;
     cctrl->code_ctrl->statics_offset = code_off;
     cctrl->statics_label->addr = bin + cctrl->code_ctrl->statics_offset;
diff --git a/c/x86_64_backend.c b/c/x86_64_backend.c
index 62c07cc..86f9c12 100644
--- a/c/x86_64_backend.c
+++ b/c/x86_64_backend.c
@@ -8649,7 +8649,8 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
       bin = MemGetWritePtr(xbin);
       code_off = 0;
     }
-    code_off = FuncProlog(cctrl, bin, code_off);
+    if (!(cctrl->flags & CCF_RAW_AOT_ASM))
+      code_off = FuncProlog(cctrl, bin, code_off);
     for (cnt = 0; cnt < cnt2; cnt++) {
     enter:
       cctrl->backend_user_data1 = 0;
@@ -8686,7 +8687,8 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
       }
     }
     cctrl->epilog_label->addr = bin + code_off;
-    code_off = FuncEpilog(cctrl, bin, code_off);
+    if (!(cctrl->flags & CCF_RAW_AOT_ASM))
+      code_off = FuncEpilog(cctrl, bin, code_off);
     for (misc = cctrl->code_ctrl->code_misc->next;
          misc != cctrl->code_ctrl->code_misc; misc = misc->base.next) {
       int64_t old_code_off = code_off;
@@ -8769,7 +8771,10 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
           code_off += 8 - code_off % 8;
         misc->addr = bin + code_off;
         if (bin)
-          *(void **)((char *)MemGetWritePtr(bin) + code_off) = &DoNothing;
+          if (cctrl->flags & CCF_AOT_COMPILE)
+            *(void **)((char *)MemGetWritePtr(bin) + code_off) = NULL;
+          else
+            *(void **)((char *)MemGetWritePtr(bin) + code_off) = &DoNothing;
         code_off += 8;
         if (run == 2) {
           *(import = A_CALLOC(sizeof(CHashImport), NULL)) = (CHashImport){
@@ -8804,7 +8809,7 @@ char *OptPassFinal(CCmpCtrl *cctrl, int64_t *res_sz, char **dbg_info,
         goto fill_in_refs;
       }
     }
-    if (code_off % 8) // Align to 8
+    if (!(cctrl->flags & CCF_RAW_AOT_ASM) && code_off % 8) // Align to 8
       code_off += 8 - code_off % 8;
     cctrl->code_ctrl->statics_offset = code_off;
     cctrl->statics_label->addr = bin + cctrl->code_ctrl->statics_offset;
-- 
2.53.0
